google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,352 @@
+    }
+    AttemptRetry(status, retry_policy_prototype_->clone(),
+                 backoff_policy_prototype_->clone());
+  }
+
+  void AttemptRetry(Status status, std::shared_ptr<RetryPolicy> retry_policy,
+                    std::shared_ptr<BackoffPolicy> backoff_policy) {
+    bool shutdown;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      shutdown_ = shutdown;
+    }
+    if (!shutdown && !retry_policy->IsExhausted() &&
+        retry_policy->OnFailure(status)) {
+      sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void> sleep_fu) {
+            this->Initialize();
+          });
+      return;
+    }
+    std::lock_guard<std::mutex> g{mu_};
+    status_promise_.set_value(kPermanentErrorStatus);
+    in_retry_loop_ = false;
+  }
+
+  void Initialize(std::shared_ptr<RetryPolicy> retry_policy,
+                  std::shared_ptr<BackoffPolicy> backoff_policy) {
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+        potential_stream;
+    future<bool> start_future;
+
+    potential_stream = stream_factory_();
+    start_future = potential_stream->Start();
+
+    auto start_future_result = start_future.then(
+        [this, potential_stream =
+                   std::move(potential_stream)](future<bool> start_future) {
+          if (!start_future.get()) {
+            {
+              std::lock_guard<std::mutex> g{this->mu_};
+              this->in_retry_loop_ = true;
+            }
+
+            this->AttemptRetry(kStartErrorStatus, retry_policy, backoff_policy);
+            return make_ready_future(
+                StatusOr<std::unique_ptr<
+                    AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>(
+                    kStartErrorStatus));
+          }
+          return this->initializer_(std::move(potential_stream));
+        });
+
+    start_future_result.then(
+        [this, retry_policy, backoff_policy](
+            future<StatusOr<std::unique_ptr<
+                AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>
+                start_initialize_future) {
+          auto start_initialize_response = start_initialize_future.get();
+
+          if (!start_initialize_response.ok()) {
+            // Start failure which already invoked retry loop, so return
+            // immediately
+            if (start_initialize_response.status() == kStartErrorStatus) {
+              return;
+            }
+            {
+              std::lock_guard<std::mutex> g{this->mu_};
+              this->in_retry_loop_ = true;
+            }
+            this->AttemptRetry(kInitializerErrorStatus, retry_policy,
+                               backoff_policy);
+            return;
+          }
+
+          std::lock_guard<std::mutex> g{this->mu_};
+          this->stream_ = std::move(*start_initialize_response);
+          if (this->read_promise_.has_value()) {
+            this->read_promise_.value().set_value(absl::optional());
```

Comment:
set_value is an upcall, and shouldn't be called with the lock held, as it may try to call back into this class.